File: solutions/r/collatz-conjecture/collatz-conjecture.R
#  collatz_step_counter takes an integer and returns the number of steps
#  required to get the number to 1 when following the rules:
#    - if number is odd, multiply with 3 and add 1
#    - if number is even, divide by 2
collatz_step_counter <- Vectorize(function(num) {
    if (num < 1) {
        stop("invalid starting value, num must be a positive integer")
    }

    steps <- 0
    while (num > 1) {
        if (num %% 2 == 0) {
            num <- num / 2
        } else {
            num <- 3 * num + 1
        }
        steps <- steps + 1
    }
    steps
})

File: solutions/go/collatz-conjecture/collatz_conjecture.go
package collatzconjecture

import "errors"

func CollatzConjecture(n int) (int, error) {
	if n < 1 {
		return 0, errors.New("invalid starting value, n must be a positive integer")
	}

	steps := 0
	for n > 1 {
		if n%2 == 0 {
			n = n / 2
		} else {
			n = 3*n + 1
		}
		steps += 1
	}

	return steps, nil
}
